### PR TITLE
emqx: upgrade to 5.1.4 and fix linkage issue

### PR DESCRIPTION
**Pull request by id**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -186,6 +186,14 @@ jobs:
         id: brew-test-bot-formulae
         working-directory: ${{ env.BOTTLES_DIR }}
 
+      - name: Debug
```
Comment (SMillerDev) [inline .github/workflows/tests.yml:189]: Please don't adjust CI.

**Review by SMillerDev — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml (hunk 1 above)
Comment (id) [reply .github/workflows/tests.yml:189]: Of course, it's temporary - I'll remove it when I find out what happens.
I could not reproduce this issue locally after initial patch.

**Comment by id** [comment]
```
$ otool -l $(brew --prefix emqx)/lib/crypto-5.1.4/priv/lib/crypto.so
Load command 10
          cmd LC_LOAD_DYLIB
      cmdsize 104
         name @loader_path/../../../../../../../../../opt/openssl@3/lib/libcrypto.3.dylib (offset 24)
   time stamp 2 Wed Dec 31 16:00:02 1969
      current version 3.0.0
compatibility version 3.0.0
```

File: Formula/emqx.rb [hunk 2]
```diff
@@ -46,12 +46,16 @@ def install
     end
     chmod "+x", prefix/"releases/#{version}/no_dot_erlang.boot"
     bin.install_symlink prefix/"releases/#{version}/no_dot_erlang.boot"
+    if OS.mac?
+      # TODO: remove this after upgrading to erlang@26
+      system "install_name_tool", "-change", "@loader_path/../../../../../../../../../opt/openssl@3/lib/libcrypto.3.dylib", Formula["openssl@3"].opt_lib/"libcrypto.3.dylib", prefix/"lib/crypto-5.1.4/priv/lib/crypto.so"
+    end
```
Comment (carlocab) [inline Formula/emqx.rb:52]: Why do you need this?

**Comment by carlocab** [comment]
> This library should be copied verbatim from erlang@25 install, but something overwrites load command adding 2 extra `../../` in the process.

Well, not quite. The issue is that `crypto.so` lives in 
```
lib/erlang/lib/crypto-5.1.4/priv/lib
```
while `emqx` copies the `crypto.so` from `erlang@25` into
```
lib/crypto-5.1.4/priv/lib
```
which is two directories shorter (and explains the extra `../../`).

You can't just relocate binaries arbitrarily and expect them to keep working.

**Comment by id** [comment]
> while `emqx` copies the `crypto.so` from `erlang@25` into

the error have been occurring before this patch.

**Comment by id** [comment]
@carlocab erlang's crypto.so uses absolute path to load libcrypto.3.dylib, while in emqx it somehow got changed to relative.

**Comment by carlocab** [comment]
> erlang's crypto.so uses absolute path to load libcrypto.3.dylib

Not in CI, it doesn't.
```
❯ otool -L "$(brew --prefix erlang@25)/lib/erlang/lib/crypto-5.1.4/priv/lib/crypto.so"
/usr/local/opt/erlang@25/lib/erlang/lib/crypto-5.1.4/priv/lib/crypto.so:
        @loader_path/../../../../../../../../../opt/openssl@3/lib/libcrypto.3.dylib (compatibility version 3.0.0, current version 3.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)

```

**Comment by id** [comment]
Interesting, for me it was showing this:
```
/opt/homebrew//Cellar/erlang@25/25.3.2.5/lib/erlang/lib/crypto-5.1.4/priv/lib/crypto.so:
        /opt/homebrew/opt/openssl@3/lib/libcrypto.3.dylib (compatibility version 3.0.0, current version 3.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.100.3)
```

**Comment by id** [comment]
different behaviour on arm/intel?

**Comment by id** [comment]
@carlocab was there some recent change in `brew` which started overwriting load path to dynamic libraries to relative?

**Comment by id** [comment]
Ok, solved it.
Please let me know guys if you like this solution or we need to do something else.

File: Formula/e/emqx.rb [hunk 3]
```diff
@@ -38,20 +37,37 @@ class Emqx < Formula
...
+        dynlib = MachO::MachOFile.new(lib/"crypto-#{crypto_vsn}/priv/lib/#{f}")
+        old_libcrypto = dynlib.linked_dylibs.find { |d| d.include?("libcrypto.3.dylib") }
+        next if old_libcrypto.nil?
+
+        dynlib.change_install_name(old_libcrypto, new_libcrypto)
+        dynlib.write("#{dynlib.filename}.1")
+        mv "#{dynlib.filename}.1", dynlib.filename
+        MachO.codesign!(dynlib.filename) if Hardware::CPU.arm?
+      end
+    end
```
Comment (carlocab) [inline Formula/e/emqx.rb:64]: Let's make this a bit simpler:
```suggestion
    return unless OS.mac?

    # ensure load path for libcrypto is correct
    crypto_vsn = Utils.safe_popen_read("erl", "-noshell", "-eval", 'io:format("~s", [crypto:version()]), halt().').strip
    libcrypto = Formula["openssl@3"].opt_lib/shared_library("libcrypto", "3")
    %w[crypto.so otp_test_engine.so].each do |f|
      dynlib = lib/"crypto-#{crypto_vsn}/priv/lib"/f
      old_libcrypto = dynlib.dynamically_linked_libraries
                            .find { |d| d.end_with?(libcrypto.basename) }
      next if old_libcrypto.nil?

      dynlib.change_install_name(old_libcrypto, libcrypto.to_s)
      MachO.codesign!(dynlib) if Hardware::CPU.arm?
    end
```
You may have to split some lines for being too long.

**Review by carlocab — COMMENTED** [review]
Thanks, one suggestion.

Also, I noticed that the build vendors a whole lot of things that Homebrew provides (e.g. `lz4`, `rocksdb`, etc.). We should change that given https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae.

File: Formula/e/emqx.rb [hunk 4]
```diff
@@ -1,9 +1,8 @@
 class Emqx < Formula
   desc "MQTT broker for IoT"
   homepage "https://www.emqx.io/"
-  # TODO: Check if we can use unversioned `erlang` at version bump
```
Comment (carlocab) [inline Formula/e/emqx.rb:4]: This comment still seems applicable.

File: Formula/e/emqx.rb (hunk 3 above)
Comment (id) [reply Formula/e/emqx.rb]: @carlocab thanks, applied the suggestion. However, the build fails now.

Locally I see this error:
```
Error: An exception occurred within a child process:
  Errno::EACCES: Permission denied @ rb_sysopen - /opt/homebrew/Cellar/emqx/5.1.4/lib/crypto-5.1.4/priv/lib/crypto.so
```
It happens on the `change_install_name` step - that's why I added write to another file and `mv` in my version.

File: Formula/e/emqx.rb [hunk 5]
```diff
@@ -38,20 +38,36 @@ class Emqx < Formula
...
     bin.install_symlink prefix/"releases/#{version}/no_dot_erlang.boot"
+    return unless OS.mac?
+
+    # ensure load path for libcrypto is correct
+    crypto_vsn = Utils.safe_popen_read("erl", "-noshell", "-eval",
+                                       'io:format("~s", [crypto:version()]), halt().').strip
+    libcrypto = Formula["openssl@3"].opt_lib/shared_library("libcrypto", "3")
+    %w[crypto.so otp_test_engine.so].each do |f|
+      dynlib = lib/"crypto-#{crypto_vsn}/priv/lib"/f
+      old_libcrypto = dynlib.dynamically_linked_libraries
```
Comment (carlocab) [inline Formula/e/emqx.rb:58]: Sorry, this should fix the build error:
```suggestion
      old_libcrypto = dynlib.dynamically_linked_libraries(resolve_variable_references: false)
```

File: Formula/e/emqx.rb [hunk 6]
```diff
@@ -38,20 +38,36 @@ class Emqx < Formula
...
+                            .find { |d| d.end_with?(libcrypto.basename) }
+      next if old_libcrypto.nil?
+
+      dynlib.change_install_name(old_libcrypto, libcrypto.to_s)
+      MachO.codesign!(dynlib) if Hardware::CPU.arm?
```
Comment (carlocab) [inline Formula/e/emqx.rb:63]:
```suggestion
      dynlib.ensure_writable do
        dynlib.change_install_name(old_libcrypto, libcrypto.to_s)
        MachO.codesign!(dynlib) if Hardware::CPU.arm?
      end
```

**Comment by id** [comment]
> I noticed that the build vendors a whole lot of things that Homebrew provides (e.g. `lz4`, `rocksdb`, etc.). We should change that given [docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae](https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae).

Thanks, brought this up with the team. It's not a trivial issue to solve, but I agree it's reasonable to expect from a library to make use of components already present in the system, if possible.

**Review by carlocab — DISMISSED** [review]
LGTM; thanks!

**Review by SMillerDev — DISMISSED** [review]

**Comment by id** [comment]
Apparently I cannot allow maintainers to edit PR because https://github.com/orgs/community/discussions/5634

**Comment by id** [comment]
--> https://github.com/Homebrew/homebrew-core/pull/140246

**Pull request closed without merging**